src/diffusers/models/unet_2d_condition.py

```diff
@@ -123,20 +124,45 @@ def __init__(
         num_class_embeds: Optional[int] = None,
         upcast_attention: bool = False,
         resnet_time_scale_shift: str = "default",
+        time_embed_dim: int = None,
+        time_embedding_type: str = "positional", # fourier, positional
+        timestep_act_2: bool = False, 
+        class_labels_dim: Optional[int] = None,
+        conv_in_kernel: int = 3,
+        conv_out_kernel: int = 3,
+        norm_out: bool = True,
+        attn1_types: Tuple[bool, None]=(None,'self','self','self'),
+        attn2_types: Tuple[bool, None] = (None, 'cross','cross','cross'),
+        downsample: Optional[Tuple[bool]] = None,
+        upsample: Optional[Tuple[bool]] = None,
+        skip_freq: str = "layer" # layer, block
     ):
         super().__init__()
 
         self.sample_size = sample_size
-        time_embed_dim = block_out_channels[0] * 4
+        self.skip_freq = skip_freq
+        if time_embed_dim is None:
+            time_embed_dim = block_out_channels[0] * 4
 
         # input
-        self.conv_in = nn.Conv2d(in_channels, block_out_channels[0], kernel_size=3, padding=(1, 1))
+        conv_in_padding = (conv_in_kernel-1)//2
+        self.conv_in = nn.Conv2d(in_channels, block_out_channels[0], kernel_size=conv_in_kernel, padding=conv_in_padding)
 
         # time
-        self.time_proj = Timesteps(block_out_channels[0], flip_sin_to_cos, freq_shift)
-        timestep_input_dim = block_out_channels[0]
-
-        self.time_embedding = TimestepEmbedding(timestep_input_dim, time_embed_dim)
+        if time_embedding_type == "fourier":
+            if time_embed_dim % 2 != 0:
+                raise ValueError(f"`time_embed_dim` should be divisible by 2, but is {time_embed_dim}.")
+            self.time_proj =  GaussianFourierProjection(time_embed_dim//2, set_W_to_weight=False,log=False, flip_sin_to_cos=flip_sin_to_cos)
+            timestep_input_dim = time_embed_dim
+        elif time_embedding_type == "positional":
+            self.time_proj = Timesteps(block_out_channels[0], flip_sin_to_cos, freq_shift)
+            timestep_input_dim = block_out_channels[0]
+        
+        self.time_embedding = TimestepEmbedding(timestep_input_dim, time_embed_dim, act_fn=act_fn, act_2=timestep_act_2)
+        
+        # class 
+        if class_labels_dim is not None:
+            self.class_proj = nn.Linear(class_labels_dim, timestep_input_dim, bias=False)
```

Comment:
should we maybe call this `time_emb_proj` because it projects the time embeddings?

Reply:
are you referring to `class_proj`? `class_proj` projects the class embeddings - I do find the naming here a little bit confusing

(this is all the layers are used in `__forward__`)
```python
        t_emb = self.time_proj(timesteps)
        t_emb = t_emb.to(dtype=self.dtype)

        if class_labels is not None and self.class_proj is not None:
            class_emb = self.class_proj(class_labels)
            class_emb = class_emb.to(dtype=self.dtype)
            if self.class_embedding is None:
                t_emb = t_emb + class_emb

        emb = self.time_embedding(t_emb)
```